### ISSUE TITLE
samples/net: the HTTP client sample app must run on QEMU x86

### ISSUE THREAD

**Issue by nashif** [zephyriot/zep-jira14#1720]
**_Reported by jiequan wu:_**

I created a prj_qemu_x86.conf file to test http_client sample app on Qemu x86.
prj_qemu_x86.conf:
```
CONFIG_NETWORKING=y                                                                                          
CONFIG_NET_TCP=y
CONFIG_RANDOM_GENERATOR=y
CONFIG_TEST_RANDOM_GENERATOR=y
CONFIG_NET_LOG=y
CONFIG_SYS_LOG_NET_BUF_LEVEL=2
CONFIG_SYS_LOG_NET_LEVEL=2
CONFIG_NET_BUF_LOG=y
CONFIG_NET_SLIP_TAP=y
CONFIG_INIT_STACKS=y
CONFIG_SYS_LOG_SHOW_COLOR=y
CONFIG_PRINTK=y
CONFIG_NET_STATISTICS=y

CONFIG_NET_NBUF_RX_COUNT=16
CONFIG_NET_NBUF_TX_COUNT=16
CONFIG_NET_NBUF_DATA_COUNT=16

CONFIG_NET_IPV6_RA_RDNSS=y
CONFIG_NET_IF_UNICAST_IPV6_ADDR_COUNT=3
CONFIG_NET_IF_MCAST_IPV6_ADDR_COUNT=2

CONFIG_STDOUT_CONSOLE=y

CONFIG_HTTP_PARSER=y

# Enable IPv6 support
CONFIG_NET_IPV6=n
# Enable IPv4 support
CONFIG_NET_IPV4=y

CONFIG_NET_SAMPLES_IP_ADDRESSES=y
CONFIG_NET_SAMPLES_MY_IPV6_ADDR="2001:db8::1"
CONFIG_NET_SAMPLES_PEER_IPV6_ADDR="2001:db8::2"
CONFIG_NET_SAMPLES_MY_IPV4_ADDR="192.0.2.1"
CONFIG_NET_SAMPLES_PEER_IPV4_ADDR="192.0.2.2"

CONFIG_NET_MAX_CONTEXTS=16

CONFIG_MBEDTLS=y
CONFIG_MBEDTLS_BUILTIN=y
CONFIG_MBEDTLS_CFG_FILE="config-mini-tls1_2.h"
```
Steps:
Open a terminal window and type:
./loop-socat.sh
Open another terminal window and type:
sudo ./loop-slip-tap.sh 
cd samples/net/http_client
make pristine && make BOARD=qemu_x86 run

Qemu output and building with some warnings:
```
/home/linux/latest/zephyr.latest/samples/net/http_client/src/http_client.c: In function ‘http_send_request’:
/home/linux/latest/zephyr.latest/samples/net/http_client/src/http_client.c:98:32: warning: format ‘%u’ expects argument of type ‘unsigned int’, but argument 4 has type ‘size_t {aka long unsigned int}’ [-Wformat=]
          "\r\nContent-Length: %u\r\n\r\n",
                                ^
  CC      src/http_client_rcv.o
  CC      src/http_client_cb.o
/home/linux/latest/zephyr.latest/samples/net/http_client/src/http_client_cb.c: In function ‘on_url’:
/home/linux/latest/zephyr.latest/samples/net/http_client/src/http_client_cb.c:19:17: warning: field precision specifier ‘.*’ expects argument of type ‘int’, but argument 2 has type ‘size_t {aka long unsigned int}’ [-Wformat=]
  printf("URL: %.*s\n", length, at);
                 ^
/home/linux/latest/zephyr.latest/samples/net/http_client/src/http_client_cb.c: In function ‘on_header_field’:
/home/linux/latest/zephyr.latest/samples/net/http_client/src/http_client_cb.c:52:12: warning: field precision specifier ‘.*’ expects argument of type ‘int’, but argument 2 has type ‘size_t {aka long unsigned int}’ [-Wformat=]
  printf("%.*s: ", length, at);
            ^
/home/linux/latest/zephyr.latest/samples/net/http_client/src/http_client_cb.c: In function ‘on_header_value’:
/home/linux/latest/zephyr.latest/samples/net/http_client/src/http_client_cb.c:81:12: warning: field precision specifier ‘.*’ expects argument of type ‘int’, but argument 2 has type ‘size_t {aka long unsigned int}’ [-Wformat=]
  printf("%.*s\n", length, at);
            ^
  LD      src/built-in.o
  AR      libzephyr.a
  LINK    zephyr.lnk
  SIDT    staticIdt.o
  LINK    zephyr.elf
  BIN     zephyr.bin
To exit from QEMU enter: 'CTRL+a, x'
[QEMU] CPU: qemu32
qemu-system-i386: warning: Unknown firmware file in legacy mode: genroms/multiboot.bin

```

(Imported from Jira ZEP-1872)

**Comment by nashif** [comment]
**_by Flavio Santes:_**

Thanks for reporting, this issue is only seen when building for QEMU and we already solve it: 

https://gerrit.zephyrproject.org/r/#/c/12045/
https://gerrit.zephyrproject.org/r/#/c/12046/
https://gerrit.zephyrproject.org/r/#/c/12047/

**Comment by nashif** [comment]
**_by Flavio Santes:_**

I am going to change this from "Bug" to "Story" because the original requirement was to support only the FRDM K64F board.

**Comment by nashif** [comment]
**_by Sharron LIU:_**

This issue also indicate sample apps configure missing. Please address the necessity to have one configure for each board.

**Comment by nashif** [comment]
**_by jiequan wu:_**

Flavio Santes, I tested it on v1.7-rc4(commit id 3d2893cf85d5).
Qemu output:
```
/home/linux/latest/zephyr.latest/samples/net/http_client/src/http_client.c: In function ‘http_send_request’:
/home/linux/latest/zephyr.latest/samples/net/http_client/src/http_client.c:98:32: warning: format ‘%u’ expects argument of type ‘unsigned int’, but argument 4 has type ‘size_t {aka long unsigned int}’ [-Wformat=]
          "\r\nContent-Length: %u\r\n\r\n",
                                ^
  CC      src/http_client_rcv.o
  CC      src/http_client_cb.o
  LD      src/built-in.o
  AR      libzephyr.a
  LINK    zephyr.lnk
  SIDT    staticIdt.o
  LINK    zephyr.elf
  BIN     zephyr.bin
To exit from QEMU enter: 'CTRL+a, x'
[QEMU] CPU: qemu32
qemu-system-i386: warning: Unknown firmware file in legacy mode: genroms/multiboot.bin
```
And the http_client seems to still does not work on Qemu.

**Comment by nashif** [comment]
**_by Flavio Santes:_**

_jiequan wu_ : patches were sent to the net branch. For 1.7-rc4, the patches must be cherry picked by the maintainer. Test again in net and let me know. Thanks.

**Comment by nashif** [comment]
**_by Kuo-Lang Tseng:_**

_jiequan wu_ , I just cherry-picked the related patches _Flavio Santes_  mentioned to 1.7 branch. Will ping to get them merged.

**Comment by nashif** [comment]
**_by Kuo-Lang Tseng:_**

https://gerrit.zephyrproject.org/r/#/c/12156/
https://gerrit.zephyrproject.org/r/#/c/12157/
https://gerrit.zephyrproject.org/r/#/c/12158/

**Comment by nashif** [comment]
**_by Kuo-Lang Tseng:_**

The above cherrpicks were merged to 1.7 branch. _jiequan wu_ , can you please verify? Thanks.

**Comment by nashif** [comment]
**_by jiequan wu:_**

I tested on master branch(commit c2cc5f90e2), it is fine.

**Comment by nashif** [comment]
**_by jiequan wu:_**

Kuo-Lang Tseng, I tested it on  v1.7.0(commit id: bdc3107d5). This issue still exist.

**Comment by nashif** [comment]
**_by jiequan wu:_**

I saw the patch: https://gerrit.zephyrproject.org/r/#/c/12158/  was missing in v1.7.0(commit id: bdc3107d5).

**Comment by nashif** [comment]
**_by Kuo-Lang Tseng:_**

_jiequan wu_ , that's right. That patch didn't get merged to 1.7 branch so it is not part of 1.7 release. Since 1.7 is already released, it's no point to keep the pending patch there so I abandoned it. Since the fix was already merged to master branch, it should work on master for future releases.
